std/bigint.d

```diff
@@ -565,12 +566,27 @@ public:
     {
         auto  x  = BigInt("1_000_000_500");
         long  l  = 1_000_000L;
-        ulong ul = 2_000_000UL;
+        ulong ul = 1_000_000UL;
         int   i  = 500_000;
         short s  = 30_000;
 
+        import std.stdio;
+        () @trusted {
+            writeln(is(typeof(x % l) == long),
+                    ", x:", x,
+                    ", l:", l,
+                    ", x%l:", x % l);
+        } ();
         assert(is(typeof(x % l)  == long)   && x % l  == 500L);
+
+        () @trusted {
+            writeln(is(typeof(x % ul) == BigInt),
+                    ", x:", x,
+                    ", ul:", ul,
+                    ", x%ul:", x % ul);
+        } ();
         assert(is(typeof(x % ul) == BigInt) && x % ul == BigInt(500));
```

Comment:
This line fails via

```
make -f posix.mak BUILD=debug std/bigint.test
```

as

```
true, x:1000000500, l:1000000, x%l:500
true, x:1000000500, ul:1000000, x%ul:0
core.exception.AssertError@std/bigint.d(588): unittest failure
```

Something is very wrong here.

Reply:
You made a `BigUint` constructor `this(scope immutable(BigDigit) [] x) pure nothrow @nogc @trusted` but it assigns `x` to `this.data`, therefore escaping the constructor's scope. I suspect this introduces memory corruption.

Reply:
Ahh. I mistook x for having infinite lifetime. My mistake. I'll fix. Thanks!

Reply:
Removing `scope` for `x` makes

```d
    BigUint opBinary(string op, Tulong)(Tulong y) pure nothrow @safe const scope
        if (op == ">>" && is (Tulong == ulong))
    {
        assert(y > 0, "Can not right shift BigUint by 0");
        uint bits = cast(uint) y & BIGDIGITSHIFTMASK;
        if ((y >> LG2BIGDIGITBITS) >= data.length) return BigUint(ZERO);
        uint words = cast(uint)(y >> LG2BIGDIGITBITS);
        if (bits == 0)
        {
            return BigUint(data[words..$]);
        }
        else
        {
            uint [] result = new BigDigit[data.length - words]; // TODO this fails
            multibyteShr(result, data[words..$], bits);

            if (result.length > 1 && result[$-1] == 0)
                return BigUint(trustedAssumeUnique(result[0 .. $-1]));
            else
                return BigUint(trustedAssumeUnique(result));
        }
    }
```

fails as

```
std/internal/math/biguintcore.d(722): Error: scope variable `this` assigned to non-scope parameter `x` calling std.internal.math.biguintcore.BigUint.this
std/bigint.d(314): Error: template instance `std.internal.math.biguintcore.BigUint.opBinary!(">>", ulong)` error instantiating
std/bigint.d(2356):        instantiated from here: `opOpAssign!(">>", int)`
make: *** [posix.mak:342: generated/linux/debug/64/libphobos2.a] Fel 1
```

. Is there another way of fixing this than replacing `data[words..$]` with `data[words..$].idup`?

I don't understand why `data[words..$]` can't be passed to the `BigUint.this(immutable(BigDigit) [] x)`.

?

Reply:
This was fixed when I removed `scope` from `opBinary`.